### PR TITLE
Add CIFAR10 Dataset

### PR DESCRIPTION
**Pull request by Delaunay**

### CODE REVIEW

**Comment by breuleux** [comment]
This overlaps #8, we need to synchronize work on this feature.

File: milatools/datasets/CIFAR10.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        splits = split(self, method)
+
+        trainset = Subset(self, splits.train)
+        validset = Subset(self, splits.valid)
+        testset = Subset(self, splits.test)
+
+        if final:
+            trainset = ConcatDataset([trainset, validset])
+            # Valid set get merged to get a bigger trainset
+            # when HPO is done
```
Comment (manuel-delverme) [inline milatools/datasets/CIFAR10.py:100]: What is HPO?

File: milatools/datasets/CIFAR10.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    test_size = 10000
+
+    @staticmethod
+    def mila_path():
+        """Path to where the dataset is stored inside the mila network"""
+        return "/network/datasets/cifar10.var/cifar10_torchvision/"
+
+    @staticmethod
+    def build_dataset(*args, **kwargs):
+        """Builds the expexted dataset"""
```
Comment (manuel-delverme) [inline milatools/datasets/CIFAR10.py:49]: typo

File: milatools/datasets/CIFAR10.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        import torchvision.datasets as datasets
+        from torch.utils.data import ConcatDataset
```
Comment (manuel-delverme) [inline milatools/datasets/CIFAR10.py:51]: is there a reason the imports are module level?

File: milatools/datasets/CIFAR10.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
+from milatools.datasets.cached import CachedDataset
```
Comment (manuel-delverme) [inline milatools/datasets/CIFAR10.py:1]: Usually the from imports are in the form `from milatools.datasets import cached` (https://google.github.io/styleguide/pyguide.html#22-imports)
to reduce namespace clashes, I'm not sure about the type style we want to have in this repo, @breuleux, what do you think?

File: milatools/datasets/CIFAR10.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+    @staticmethod
+    def build_dataset(*args, **kwargs):
+        """Builds the expexted dataset"""
+        import torchvision.datasets as datasets
+        from torch.utils.data import ConcatDataset
+
+        train_dataset = datasets.CIFAR10(*args, train=True, **kwargs)
+        test_dataset = datasets.CIFAR10(*args, train=False, **kwargs)
+        return ConcatDataset([train_dataset, test_dataset])
```
Comment (manuel-delverme) [inline milatools/datasets/CIFAR10.py:55]: Is having a concatenated dataset (and len(dataset) = train + test) standard?

in pytorch examples they have two different datasets and threats them independently:
https://github.com/pytorch/examples/blob/master/imagenet/main.py#L208

Having said that I'm not doing that much supervised learning so I might be missing some good reason for this.

File: milatools/datasets/split.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
+from dataclasses import dataclass
+
+import numpy
+
+
+@dataclass
+class Split:
+    """Returns 3 Splits of the main data set. The splits are indices of the samples inside the data set"""
+
+    train: numpy.array
```
Comment (manuel-delverme) [inline milatools/datasets/split.py:10]: was this meant to be torch.Tensor? these are loaded from the torchvision location anyway

File: milatools/datasets/split.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+    def keys(self):
+        return "train", "valid", "test"
+
+    def values(self):
+        return self.train, self.valid, self.test
+
+
+def original_split(
+    datasets, data_size=None, seed=None, ratio=None, index=None, balanced=None
```
Comment (manuel-delverme) [inline milatools/datasets/split.py:31]: All the parameters seem unused?

**Review by manuel-delverme — COMMENTED** [review]
Thank you for the implementation!
I've skimmed it to understand the overall structure, I've left some questions around the code.

File: milatools/datasets/CIFAR10.py (hunk 1 above)
Comment (Delaunay) [reply milatools/datasets/CIFAR10.py:100]: Hyper parameter Optimization

File: milatools/datasets/CIFAR10.py (hunk 3 above)
Comment (Delaunay) [reply milatools/datasets/CIFAR10.py:51]: Wanted to avoid a hard dependency on pytorch

File: milatools/datasets/CIFAR10.py (hunk 4 above)
Comment (Delaunay) [reply milatools/datasets/CIFAR10.py:1]: never had an issue with it, you can always rename with 'as' it in case of clashes

File: milatools/datasets/CIFAR10.py (hunk 5 above)
Comment (Delaunay) [reply milatools/datasets/CIFAR10.py:55]: It is a slightly different from what people tend to use, but this is to make it easier to use randomized splits which should be used.
Since using the same splits all the time creates a bias towards models that work well on specific splits.
I linked a paper on why it is important.

While I agree that not mirroring pytorch API is not great here it is done to favour best practices.
Generating split dataset from the ConcatDataset is trivial

File: milatools/datasets/split.py (hunk 6 above)
Comment (Delaunay) [reply milatools/datasets/split.py:10]: no, those are indices not the actual sample, those a simple 1-dim array

File: milatools/datasets/split.py (hunk 7 above)
Comment (Delaunay) [reply milatools/datasets/split.py:31]: I did not yet put the [other split methods](https://github.com/mila-iqia/olympus/tree/master/olympus/datasets/split) but there will be used

File: milatools/datasets/cached.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    valid_size = None
+    test_size = None
+
+    def __init__(self, root, download=False):
+        if os.path.exists(self.mila_path()):
+            download = False
+
+        if not CachedDataset.WAS_COPIED:
+            shutil.copytree(self.mila_path(), root, dirs_exist_ok=True)
+            CachedDataset.WAS_COPIED = True
```
[the remark this answers is not in the document]
Comment (Delaunay) [reply milatools/datasets/cached.py:19]: We can use [Value][1] for multiprocessing. 
Yeah, it might be an issue for multiple datasets

[1]: https://www.kite.com/python/docs/multiprocessing.Value

File: milatools/datasets/split.py (hunk 7 above)
Comment (manuel-delverme) [reply milatools/datasets/split.py:31]: I can't see that repo :cry:

File: milatools/datasets/CIFAR10.py (hunk 3 above)
Comment (manuel-delverme) [reply milatools/datasets/CIFAR10.py:51]: We might not add it to the dependency list (i think we should) but as an user I'd rather know I'm importing something unusable rather than successfully use something that is guaranteed to fail later.

Comment (breuleux) [reply milatools/datasets/CIFAR10.py:51]: PyTorch is not the only framework, and even though it is the most popular now, that might not remain the case in a few years, so it's best to avoid a hard dependency.

This being said, if we organize it so that torch datasets are imported through `milatools.torch.datasets`, it would be fine to do the import eagerly in `milatools.torch`. We can also offer a torch "extra" that can be installed as `pip install milatools[torch]`.

File: milatools/datasets/cached.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Delaunay) [reply milatools/datasets/cached.py:19]: Actually with the distributed module we can use has_dataset_authority() to know if the worker should be downloading it or not. That is how pytorch usually does it

[the remark this answers is not in the document]
Comment (Delaunay) [reply milatools/datasets/cached.py:19]: So `WAS_COPIED` is going to be fine, it is not shared.

```
python
Python 3.8.3 (default, May 19 2020, 18:47:26) 
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> class A:
...     i = 2
... 
>>> class B(A):
...     pass
... 
>>> A.i
2
>>> B.i
2
>>> B.i = 3
>>> A.i
2
```

[the remark this answers is not in the document]
Comment (breuleux) [reply milatools/datasets/cached.py:19]: My point is that we're not setting the attribute on the subclasses, the code always reads or writes the attribute on `CachedDataset`. To have it work on a per class basis we would need to use `type(self).WAS_COPIED`.

File: milatools/datasets/CIFAR10.py (hunk 3 above)
Comment (Delaunay) [reply milatools/datasets/CIFAR10.py]: Moved everything to a torch modile

File: milatools/datasets/split.py (hunk 7 above)
Comment (Delaunay) [reply milatools/datasets/split.py]: I added one of the split as an example

File: milatools/datasets/cached.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Delaunay) [reply milatools/datasets/cached.py]: I did the modification

**Comment by lebrice** [comment]
I suggest we close this, wdyt @breuleux ?

**Comment by breuleux** [comment]
Indeed, this was part of the original vision but it is now out of scope. Milatools is the library to install on the user's machine and libraries that should be run on the cluster will be in different projects (see @satyaog's comment above).
